DSpace/DSpace
# Facet search should search all words, not just the first one

## References
* Fixes Dspace/dspace-angular#1712

## Description
Facets now also get indexed with `fieldName_prefix` fields. The value is stored starting from each word until the end, together with the original value. E.g. subject "With Multiple Words" gets stored as: 
```
subject_prefix =  with multiple words ||| With Multiple Words,
                  multiple words ||| With Multiple Words,
                  words ||| With Multiple Words
```
This makes it so the facet search box will search for all words, instead of just the first. E.g. searching for "John" or "Smith" in the author search box will both result in `Smith, John`.

**Note**: Word splitting is done by regex word boundaries. This means not only spaces are used to delineate values, but also characters like `- , ' : | ;`.

## Instructions for Reviewers
### Changes
- ItemIndexFactoryImpl:
  - Put facet indexing in seperate `indexIfFilterTypeFacet` method.
  - Also add `_prefix` fields for filters of type TEXT and HIERARCHICAL.
  - Recursively determine value of this field with `saveFacetPrefixParts` method.
- SolrServiceImpl:
  - New `transformPrefixFacetField` method which is only used in case of a prefix search.
  - In case of a prefix search, `_prefix` field now gets used, instead of `_filter` field.
- DiscoverResult:
  - Minor change to make conversion with `_prefix` fields work.
- schema.xml
  - Added `_prefix` field to fields used for sidebar filters & Solr browse by value.
- DiscoveryRestControllerIT
  - Added ITs to prefix search for multiple words, both for author and subject, both with/without authorityKey.
- FacetValueMatcher
  - Minor change to fix subject matchers.
- SolrServiceMetadataBrowseIndexingPlugin
  - `_prefix` field also needs to be added for browsing. This is without the "starting from each word until the end"-part however. So the `_prefix` field for browsing is essentially just a copy of the `_filter` field.
 
**Note on indexing for Browse by**: Currently the search terms used for the prefix filter/search in `{ui.url}/browse/{browseName}` ([example](https://demo7.dspace.org/browse/author?bbm.page=1&startsWith=test) | endpoint `/api/discover/browses/{browseName}/entries?startsWith={query}`) also only work on the first 'word' of the browse entry value. This work/PR could likely be expanded to have the same search functionality to search on each 'word' for the browse, by populating the `bi_*_prefix` field with the same value splitting logic as done for the search sidebar facets. But this wasn't in scope of the requested feature of [#1712](https://github.com/DSpace/dspace-angular/issues/1712), so current functionality remains unchanged. 

### How to test:

**Note**: `search/schema.xml` has been altered, so it needs to be adjusted and restarted. Discovery re-index is also necessary.

Add some items with author "Smith, John" and subject "Test:with:colon"; author "d'Arc, Joan" and subject "Test,with,comma"; author "Anne-marie" and subject "Test title;with;semicolon".

Go to the general search page on DSpace, use the Author sidebar search box and verify: 
- `Smith` suggests "Smith, John" (as before).
- `jo` suggests "Smith, John" and "d'Arc, Joan".
- `ar` suggests "d'Arc, Joan".
- `mARie` suggests "Anne-marie".

Go to the general search page on DSpace, use the Subject sidebar search box and verify: 
- `Test` suggests all 3 subjects (as before).
- `CO` suggests "Test,with,comma" and "Test:with:colon".
- `title` suggests "Test title;with;semicolon".
- `seMi` suggests "Test title;with;semicolon".

Facet endpoints:
- `/server/api/discover/facets/author?prefix=j` should return "Smith, John" and "d'Arc, Joan".
- `/server/api/discover/facets/subject?prefix=WITH` should return all 3 subjects.

Searching for browse entries should still work as before. Go to the `/browse/author` page and verify: 
- `Smith` returns "Smith, John".
- `John` returns nothing.

Browse endpoints:
- `/server/api/discover/browses/author/entries?startsWith=S` should return "Smith, John".
- `/server/api/discover/browses/author/entries?startsWith=J` should return nothing.

## Checklist
